pylint/checkers/modified_iteration_checker.py

```diff
@@ -0,0 +1,132 @@
+        ),
+        "E4714": (
+            "Iterated dict '%s' is being modified inside loop body, iterate through a copy of it instead.",
+            "iterating-modified-dict",
+            "Emitted when items are added or removed to a dict being iterated through."
+            "Doing so raises a RuntimeError",
+        ),
+        "E4715": (
+            "Iterated set '%s' is being modified inside loop body, iterate through a copy of it instead.",
+            "iterating-modified-set",
+            "Emitted when items are added or removed to a set being iterated through."
+            "Doing so raises a RuntimeError",
+        ),
+    }
+
+    options = ()
+    priority = -2
+
+    @utils.check_messages(
+        "iterating-modified-list", "iterating-modified-dict", "iterating-modified-set"
+    )
+    def visit_for(self, node: nodes.For) -> None:
+        iter_obj = node.iter
+        for body_node in node.body:
+            if self._iterating_modified_list_cond(body_node, iter_obj):
+                self.add_message(
+                    "iterating-modified-list",
+                    node=body_node,
+                    args=(iter_obj.name,),
+                )
+            elif self._iterating_modified_dict_cond(body_node, iter_obj):
+                self.add_message(
+                    "iterating-modified-dict",
+                    node=body_node,
+                    args=(iter_obj.name,),
+                )
+            elif self._iterating_modified_set_cond(body_node, iter_obj):
+                self.add_message(
+                    "iterating-modified-set",
+                    node=body_node,
+                    args=(iter_obj.name,),
+                )
+
+    @staticmethod
+    def _is_attribute_call_expr(body_node) -> bool:
+        return (
+            isinstance(body_node, astroid.Expr)
+            and isinstance(body_node.value, astroid.Call)
+            and isinstance(body_node.value.func, astroid.Attribute)
+        )
+
+    @classmethod
+    def _common_cond_list_set(cls, node, list_obj, infer_val):
+        return (
+            cls._is_attribute_call_expr(node)
+            and (infer_val == list_obj.inferred()[0])
+            and (node.value.func.expr.name == list_obj.name)
+        )
+
+    @staticmethod
+    def _dict_node_cond(node):
+        return isinstance(node, astroid.Assign) and (
+            isinstance(node.targets[0], astroid.Subscript)
+        )
+
+    @classmethod
+    def _iterating_modified_list_cond(cls, node, list_obj):
+        try:
+            infer_val = next(node.value.func.expr.infer())
+            return (
+                cls._common_cond_list_set(node, list_obj, infer_val)
+                and (infer_val.pytype() == "builtins.list")
+                and node.value.func.attrname in {"append", "remove"}
+            )
+        except (astroid.InferenceError, AttributeError):
```

Comment:
I'd like for these exceptions to be caught by the correct `isinstance` checks rather than catching them all and returning `False`. We don't know what is failing here and this might actually be exposing real bugs in `pylint` or `astroid`.
Same goes for the others.

Reply:
okay  I will move the `isinstance`  conditions before any infer, so we try to infer only when we are on the correct expression type